aten/src/ATen/native/Distance.cpp

```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+Tensor to_squareform(const Tensor& self) 
+{
+  AT_CHECK(self.dim() == 1,
+    "expected vector-form distance as input, dim as 1, but got ", self.dim());
+  int64_t size = self.size(0), d = (std::sqrt(8 * size + 1) + 1) / 2;
+  AT_CHECK(d * (d - 1) / 2 == size,
+    "not able to derive dimension, d from size ", size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
+  for (int64_t i = 0, idx = 0; i < d; i++)
+  {
+    for (int64_t j = i + 1; j < d && idx < size; j++)
+    {
+      out_tensor[i][j] = self[idx];
```

Comment:
I think this will be slow because you create multiple tensors each time through the loop (can you benchmark against numpy?).  Instead you could operate on .data and use the strides.